File: c++/1557.cpp
// Link: https://leetcode.com/problems/minimum-number-of-vertices-to-reach-all-nodes/description/
// Runtime: 312 ms
// Memory: 94.1 MB

class Solution 
{
public:
    std::vector<int> findSmallestSetOfVertices(int n,
                                               const std::vector<std::vector<int>>& edges) 
    {
        std::vector<int> waysToNodes(n);
        for (const auto& edge : edges)
            waysToNodes[edge[1]]++;
        
        std::vector<int> unreachableNodes;
        for (int i = 0; i < n; i++)
            if (waysToNodes[i] == 0)
                unreachableNodes.push_back(i);
        
        return unreachableNodes;
    }
};

File: python/1557.py
# Link: https://leetcode.com/problems/minimum-number-of-vertices-to-reach-all-nodes/description/
# Runtime: 1157 ms
# Memory: 55 MB


class Solution:
    def findSmallestSetOfVertices(self, n: int, edges: List[List[int]]) -> List[int]:
        waysToNodes = [0] * n
        for edge in edges:
            waysToNodes[edge[1]] += 1

        unreachableNodes = []
        for index, value in enumerate(waysToNodes):
            if value == 0:
                unreachableNodes.append(index)

        return unreachableNodes
